src/main/java/sparksoniq/spark/iterator/flowr/WhereClauseSparkIterator.java

```diff
@@ -20,30 +20,103 @@
+import javax.naming.OperationNotSupportedException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class WhereClauseSparkIterator extends SparkRuntimeTupleIterator {
+
+    private RuntimeIterator _expression;
+    private DynamicContext _tupleContext;   // re-use same DynamicContext object for efficiency
+    private FlworTuple _nextLocalTupleResult;
+    private FlworTuple _inputTuple;     // tuple received from child, used for tuple creation
+
     public WhereClauseSparkIterator(RuntimeTupleIterator child, RuntimeIterator whereExpression, IteratorMetadata iteratorMetadata) {
-        super(child, null, FLWOR_CLAUSES.WHERE, iteratorMetadata);
-        this._children.add(whereExpression);
+        super(child, iteratorMetadata);
+        _expression = whereExpression;
+    }
+
+    @Override
+    public boolean isRDD() {
+        return _child.isRDD();
+    }
+
+    @Override
+    public void open(DynamicContext context) {
+        super.open(context);
+
+        // isRDD checks omitted, as open is used for non-RDD(local) operations
+
+        if (this._child != null) {
+            _child.open(_currentDynamicContext);
+            _tupleContext = new DynamicContext(_currentDynamicContext);     // assign current context as parent
+
+            setNextLocalTupleResult();
+
+        } else {
+            throw new SparksoniqRuntimeException("Invalid where clause.");
+        }
+    }
+
+    @Override
+    public FlworTuple next() {
+        if(_hasNext == true){
+            FlworTuple result = _nextLocalTupleResult;      // save the result to be returned
+            setNextLocalTupleResult();              // calculate and store the next result
+            return result;
+        }
+        throw new IteratorFlowException("Invalid next() call in let flwor clause", getMetadata());
+    }
+
+    private void setNextLocalTupleResult() {
+        // for each incoming tuple, evaluate the expression to a boolean.
+        // forward if true, drop if false
+
+        while (_child.hasNext()) {
+            _inputTuple = _child.next();
+            _tupleContext.removeAllVariables();             // clear the previous variables
+            _tupleContext.setBindingsFromTuple(_inputTuple);      // assign new variables from new tuple
+
+            _expression.open(_tupleContext);
+            Item result = _expression.next();
+            _expression.close();
+            try {
+                if (result.getBooleanValue()) {
+                    _nextLocalTupleResult = _inputTuple;
+                    this._hasNext = true;
+                    return;
+                }
+            } catch (OperationNotSupportedException e) {
+                e.printStackTrace();
+            }
+        }
+
+        // execution reaches here when there are no more results
+        _child.close();
+        this._hasNext = false;
     }
 
     @Override
     public JavaRDD<FlworTuple> getRDD() {
         if (this._rdd == null) {
```

Comment:
As discussed before, the RDD should not be cached. But this can also be handled in your other PR in which you change the signature of this function.